text/3735-field-projections.md

```diff
@@ -0,0 +1,1348 @@
+    `.project()` and then receives a type with each field replaced with the respective projected
+    field.
+  - [cell-project] provides cell projection via a macro at the projection-site: the user writes
+    `cell_project!($ty, $val.$field)` where `$ty` is the type of `$val`. Internally, it uses unsafe
+    to facilitate the projection.
+  - [pin-projections] provides pin projections, it differs from [`pin-project`] by providing
+    explicit projection functions for each field. It also can generate other types of getters for
+    fields. [`pin-project`] seems like a more mature solution.
+- `&[mut] MaybeUninit<T>` projections
+  - [project-uninit] provides uninit projections via macros at the projection-site uses `unsafe`
+    internally.
+- multiple of the above
+  - [`field-project`] provides projection for `Pin<&[mut] T>` and `&[mut] MaybeUninit<T>` via a macro
+    at the projection-site: the user writes `proj!($var.$field)` to project to `$field`.
+  - [`field-projection`] is an experimental crate that implements general field projections via a
+    proc-macro that hashes the name of the field to create unique types for each field that can then
+    implement traits to make different output types for projections.
+
+[`field-project`]: https://crates.io/crates/field-project
+[`cell-project`]: https://crates.io/crates/cell-project
+[`pin-projections`]: https://crates.io/crates/pin-projections
+[`project-uninit`]: https://crates.io/crates/project-uninit
+[`field-projection`]: https://crates.io/crates/field-projection
+
+## Blog Posts and Discussions
+
+- [Safe Cell field projection in
+  Rust](https://www.abubalay.com/blog/2020/01/05/cell-field-projection)
+- [Field projdection for `Rc` and
+  `Arc`](https://internals.rust-lang.org/t/field-projection-for-rc-and-arc/15827)
+- [Generic Field Projection](https://internals.rust-lang.org/t/generic-field-projection/16204)
+
+Blog posts about pin (projections):
+- [Pinned places](https://without.boats/blog/pinned-places/)
+- [Overwrite trait](https://smallcultfollowing.com/babysteps/series/overwrite-trait/)
+
+## Rust and Other Languages
+
+Rust already has a precedent for compiler-generated types. All functions and closures have a unique,
+unnameable type.
+
+In C++ there are field projections supported on `std::shared_ptr`, it consists of two pointers, one
+pointing to the reference count and the other to the data. Making it possible to project down to a
+field and still take a reference count on the entire struct, keeping also the field alive.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Syntax Bikeshedding
+
+What is the right syntax for the various operations given in this RFC?
+
+Ideally we would have a strong opinion when this feature is implemented. But the decision should
+only be finalized when stabilizing the feature.
+
+### Field Projection Operator
+
+Current favorite: `$base:expr->$field:ident`.
+
+Alternatives:
+- use `~` instead of `->`
```

Comment:
for `field: T`, the expression `base->field` producing any type other than a place of type `T` would be very confusing for readers knowing C and C++.

Reply:
I think someone coming from C might find `&raw const base->field` most familiar while still being in line with the difference between `&` and `&raw const`. I'm not sure that's much more ergonomic than `&raw const (*base).field` though (except that it could be safe which is still a huge improvement).

Reply:
There are a lot of things in Rust that are very confusing for readers that know C and C++. I personally don't think this is a huge issue, because for pointer types the operation is very similar to "getting a place" and for types it is just a generalization.

That said, I don't really mind the exact syntax of the operator as long as it's short and conveys the correct meaning. What would be your suggestion for an alternative?
Do you want me to add anything to the RFC (e.g. add this as a drawback)?

Reply:
yeah I guess adding `&raw const` really would only be immediately obvious for raw pointers anyway. I agree it should be concise and uniform across containers so I think `->` might be the best option. If it's introduced in the context of another container like `MaybeUninit` or `Pin` `->` shouldn't be an issue.

Reply:
Both `~` and `->` are fine with me. Just a note, in the previous RFC, [it was argued](https://github.com/rust-lang/rfcs/pull/3318#issuecomment-1264325053) that "[t]he ~ symbol was removed from Rust because it is hard to type on a number of keyboard layouts".

Reply:
I don't buy that `~` is removed because of AltGr, the quoted French and German keyboard layouts showed `[]{}\|` all involved AltGr yet we have no problem using any of these 6 symbols. Additionally there is that `X: ~const Trait` syntax. I think it's more that we don't want to waste the remaining unused symbols.

Reply:
> Additionally there is that X: ~const Trait syntax.

That's placeholder syntax, don't use it as precedent please.

Reply:
I find `~` just looks very odd, so I personally slightly prefer `->`. It does something with fields behind pointers, so IMO it's "close enough" to the C/C++ equivalent.

But in the end it's just syntax, I don't care very strongly and find it hard to predict how others will react.

Reply:
> I would personally advocate `~` here, because it feels like it wouldn't have the C++ connotations of `->`, and because it is more of a "navigate" operation than a "dereference" operation.

I don't think the C++ bit holds when `~` is used for bitwise not and destructors, field projection seems more akin to `->` member access than either of those :)